### PR TITLE
WIP: GeoJSON Data Source

### PR DESCRIPTION
**Pull request by birdsarah**
WIP Currently this only implements Points in a Feature Collection and still missing lots of things esp on the JS side. Will continue updating, but feel free to add feedback comments on approach.

Will add a check list of long list of things left to do on Monday.

### CODE REVIEW

**Comment by bryevdv** [comment]
@birdsarah this is quite a bit different from what I had in mind, which is not necessarily bad, but maybe we should have a discussion about design so we are on the same page? I definitely did not have in mind that a `GeoJSONDataSource` would be a subclass of a `ColumnDataSource`. There are lots of things that GeoJSON supports that don't really fit well at all in the "tabular data" abstraction, and I think shoe-horning GeoJSON data into a table is mistake, even if we try to hide it and do it for use user automatically. I had in mind a completely new data source that accepted GeoJSON without doing any kind of adaptation, and then either:
* a completely new `GeoJSONRenderer` that operated directly on the raw GeoJSON data to render what ever it describes

or 

* adapters/transforms that can be used to hook coordinates of existing glyph renderers to the raw GeoJSON in the `GeoJSONDataSource` 

or possibly both. 

In the latter case, we really do need the ability to pull different glyph properties from different sources. Which is also fine, but that is a non-trivial thing to do the right way, and we definitely should start off by figuring and doing it the right way.

**Comment by bryevdv** [comment]
Very rough sketch:
```
table = ColumnDataSource()   # columns "foo", "bar"

geo = GeoJSONDataSource()   # provides "lat", "lon"

composite = SyntheticDataProvider(dict(x=geo, y=geo, foo=table, bar=table)) 

p.circle(x="lon", y="lat", color="foo", alpha="bar", source=composite) 
```

File: bokeh/models/sources.py [hunk 1]
```diff
@@ -254,6 +257,91 @@ def _check_column_lengths(self):
         if len(lengths) > 1:
             return str(self)
 
+
+class GeoJSONDataSource(ColumnDataSource):
+    geojson = String(help="""
+    A string of GeoJSON that contains features for plotting. Currently GeoJSONDataSource can
```
Comment (bryevdv) [inline bokeh/models/sources.py]: Should define a `GeoJSONProperty` that subclasses `String` and does validation that the string is correct `GeoJSON`. Also if we currently only support subset, that is the place to check and warn about "extra" things that we don't yet support.

**Comment by bryevdv** [comment]
Also to focus things, I think it should be an explicit design goal that the Python definition of the data source looks basically like this:
```
class GeoJSONDataSource(DataSource):

    data = GeoJSON(help="""
    A string of GeoJSON that contains features for plotting. 
    """)
```
And we should send the GeoJSON as-is to the JS side. Anything else is unnecessarily restricting, right? We may only support points but maybe users want to write some custom JS that uses other libraries and does use the rest of the data. We should not force a policy that requires throwing everything but points away.

**Comment by bryevdv** [comment]
Another reason not to do all at the adaptation on the python side is that not-doing it on the BokehJS side means other language bindings have to do alot of work to make use of the `GeoJSONDataSource`, work that just repeats what the python does and is more places for bugs.

**Comment by birdsarah** [comment]
hi @bryevdv, I look forward to chatting. I think we've got a few confusions, and I'm sorry about that, I shouldn't have put the PR up without, at least, the to check list. And the name was perhaps misleading. A GeoJOSNRenderer is necessary, but a GeoJSONDataSource (of some kind) is also necessary and seemed to be able to stand alone as a PR. Inheriting from ColumnDataSource gave a lot of bang for the buck, but yes, there are issues.

====
*Some things where I think we're confused*

> We should send the GeoJSON as-is to the JS side. 

That's exactly what's I've done.

> We may only support points but maybe users want to write some custom JS that uses other libraries and does use the rest of the data.

That's not my intention, my intention is to support all GeoJSON types I was going to finish this on Monday

> Another reason not to do all at the adaptation on the python side is that not-doing it on the BokehJS side means other language bindings have to do alot of work to make use of the GeoJSONDataSource

My intention is to do it all BokehJS side, doing it on the python side too is optional but has benefits - will allow you to interact more pythonically with the data and keeps everything in sync all the time

> `geo = GeoJSONDataSource()   # provides "lat", "lon"`

We do also need to support GeoJSON that contains property fields I believe.

=====
*Other things*

````python
# subclass bokeh.properties.JSON
class GeoJSON(JSON):
    # validation for correct GeoJSON
    # warnings about unsupported features?
````
Yes, this would be great.

>  There are lots of things that GeoJSON supports that don't really fit well at all in the "tabular data" abstraction, and I think shoe-horning GeoJSON data into a table is mistake, even if we try to hide it and do it for use user automatically.

Look forward to discussing. I agree JSON and Tabular are completely different mindsets. That being said GeoJSON is a very limited specification. The main thing I liked about this once I got started is that you get a lot of bang for your buck with not very much code.

> a completely new GeoJSONRenderer that operated directly on the raw GeoJSON data to render what ever it describes

Yes, GeoJSONRenderer was already planned as part of this or an accompanying PR, but I can see a number of other directions to the one I planned. Look forward to discussing.


> I definitely did not have in mind that a GeoJSONDataSource would be a subclass of a ColumnDataSource

I didn't either at first, but it made sense to me once I started messing around - am happy to discuss more. And also there may be sensible narrowings of this like GeoJSONColumnDataSource (probably not that name) - a data source that gives you easy wins with Bokeh but puts some constraints on your GeoJSON.

**Comment by birdsarah** [comment]
For discussion tomorrow:

Here's the meat of this GeoJSONDataSource with all geometries supported: https://github.com/bokeh/bokeh/blob/scratch/geojson_data_source/bokehjs/src/coffee/source/geojson_data_source.coffee

Here's an example: https://github.com/bokeh/bokeh/blob/scratch/geojson_data_source/examples/plotting/file/geojson_us_states.py

Here's rendered output: https://rawgit.com/bokeh/bokeh/scratch/geojson_data_source/examples/plotting/file/us_states.html

**Comment by bryevdv** [comment]
Ok it's possible I have misinterpreted or jumped to some conclusions. We can hash things out during a call. But I do feel strongly about this:

> I didn't either at first, but it made sense to me once I started messing around - am happy to discuss more. And also there may be sensible narrowings of this like GeoJSONColumnDataSource (probably not that name) - a data source that gives you easy wins with Bokeh but puts some constraints on your GeoJSON.

The main thing is I don't think there should be *any* data processing on the python side, as is done in this PR. We have made a few concessions to convenience methods on `Plot` (e.g., `add_layout`, etc.) but those are much simpler and I would say of a different character (i.e. they are shortcuts to something completely possible to do by hand). In general, all python models should be *completely declarative*. This is what allows other languages to get up and running much more quickly. So none of the adaptation to columns should happen on on the python side.

**Comment by birdsarah** [comment]
Just to stress again, the adaptation of columns that makes them usable on the JS side is not happening python side, only geojson goes down the pipe, and none of the python processing is necessary. I added it as a convenience function. 

The motivating idea was that we have lots of examples where we tweak the column data. I wanted the geojson to always be the representation and medium of transfer of the data, but allowing you a python-y way to modify the data and then at any point you could always call `source.geojson` and see/save/share the new geojson.

I was imagining examples like this:

### Static processing

```python
countries_and_boundaries = """" some geojson """"
c_b_source = GeoJSONDataSource(geojson=countries_and_boundaries)
c_b_df = c_b_source.to_df()
countries_df = c_b_df[[(cb_df.geometry_type == 'Polygon') or (cb_df.geometry_type == 'MultiPolygon')]]
boundaries_df = c_b_df[[(cb_df.geometry_type == 'LineString') or (cb_df.geometry_type == 'MultiLineString')]]
countries_source = GeoJSONDataSource(countries_df)  # note this doesn't work right now
boundaries_source = GeoJSONDataSource(boundaries_df) # note this doesn't work right now
```
So now I've written this out I don't think it's that useful, at least not yet, and there'd need to be even more work python side to make that happen.....this was a notion of a "cheap" version of a feature I want to put on the GeoJSONRenderer (more on that later).

### Server processing
```python
states = """" some geojson """"
source = GeoJSONDataSource(geojson=states)

def on_change(attr, old, new):
    source.data['line_alpha'] = [new] * len(source.data['line_alpha'])  # note this doesn't work right now

slider = Slider(start=0, end=1, value=0.5, step=0.1, title="Boundary Alpha")
slider.on_change('value', on_change)
```

---
Now I've written out the examples I'm:
a) seeing the additional work that would be needed to do them.
b) feeling that that work is, at best, a distraction

I still think there's a nugget of a good idea in my motivation for all the python side stuff, but am happy to take it out of this PR.

**Comment by birdsarah** [comment]
Am going to open new clearer PRs!

**Pull request closed without merging**